classes/observers/listeners/VersionPublicationDois.php

```diff
@@ -0,0 +1,107 @@
+
+use APP\core\Application;
+use APP\facades\Repo;
+use APP\publication\Publication;
+use Illuminate\Events\Dispatcher;
+use Illuminate\Support\LazyCollection;
+use PKP\context\Context;
+use PKP\context\ContextDAO;
+use PKP\galley\Galley;
+use PKP\observers\events\PublishedEvent;
+
+class VersionPublicationDois
+{
+    public function subscribe(Dispatcher $events): void
+    {
+        $events->listen(
+            PublishedEvent::class,
+            self::class . '@handlePublishedEvent'
+        );
+    }
+
+    /**
+     * Checks if new publication version should carry over existing DOIs or mint new ones and creates new DOIs as necessary
+     *
+     * NB: The existing doiId should be removed from the latest publication when it matches that of the previous publication,
+     * i.e. it is a copy of the previous publication. This prevents a legitimate DOI from being removed,
+     * e.g. if a versioned publication is unpublished then republished.
+     *
+     */
+    public function handlePublishedEvent(PublishedEvent $event): void
+    {
+        $submission = $event->submission;
+        $publication = $event->newPublication;
+
+        /** @var ContextDAO $contextDao */
+        $contextDao = Application::getContextDAO();
+        /** @var Context $context */
+        $context = $contextDao->getById($submission->getData('contextId'));
+
+        $shouldVersionPublication = $context->getData(Context::SETTING_DOI_VERSIONING);
+        $notFirstPublication = $publication->getData('version') != 1;
+        if ($shouldVersionPublication && $notFirstPublication) {
+
+            // Remove existing doiIds if applicable, first for publications
+            /** @var LazyCollection $publicationsSorted */
+            $publicationsSorted = $submission->getData('publications')
+                ->sortByDesc(function (Publication $publication) {
+                    return $publication->getData('version');
+                })
+                ->values();
+            /** @var Publication $latestPublication */
+            $latestPublication = $publicationsSorted->get(0);
+            /** @var Publication $previousPublication */
+            $previousPublication = $publicationsSorted->get(1);
+
+            // The existing doiId should be removed from the latest publication when it matches that of the previous publication,
+            // i.e. it is a copy of the previous publication. This prevents a legitimate DOI from being removed,
+            // e.g. if a versioned publication is unpublished then republished.
+            $hasPreviousPublicationDoiId = $latestPublication->getData('doiId') === $previousPublication->getData('doiId') && $latestPublication->getData('doiId') !== null;
+            if ($context->isDoiTypeEnabled(Repo::doi()::TYPE_PUBLICATION) && $hasPreviousPublicationDoiId) {
+                $publication->setData('doiId', null);
+                Repo::publication()->dao->update($publication);
+            }
```

Comment:
Lines 62-81 would be good to move into a protected function. I'm not sure I understand all of the inline comments and one clear, succinct docblock above a function might make it easier to understand the general intent of what's happening.

Reply:
Similar with 83-101 below.